xarray/tests/test_quantities.py

```diff
@@ -0,0 +1,160 @@
+import numpy as np
+import pytest
+
+from xarray import DataArray, set_options
+
+try:
+    import quantities as pq
+
+    has_quantities = True
+except ImportError:
+    has_quantities = False
+
+pytestmark = pytest.mark.skipif(not has_quantities, reason="requires python-quantities")
+
+
+set_options(enable_experimental_ndarray_subclass_support=True)
+
+
+def assert_equal_with_units(a, b):
+    a = a if not isinstance(a, DataArray) else a.data
+    b = b if not isinstance(b, DataArray) else b.data
+
+    assert (hasattr(a, "units") and hasattr(b, "units")) and a.units == b.units
+
+    assert (hasattr(a, "magnitude") and hasattr(b, "magnitude")) and np.allclose(
+        a.magnitude, b.magnitude
+    )
+
+
+def create_data():
+    return (np.arange(10 * 20).reshape(10, 20) + 1) * pq.V
+
+
+def create_coord_arrays():
+    x = (np.arange(10) + 1) * pq.A
+    y = np.arange(20) + 1
+    xp = (np.arange(10) + 1) * pq.J
+    return x, y, xp
+
+
+def create_coords(x=None, y=None, xp=None):
```

Comment:
You write these in whichever format works best for you; don't take this as a dictum to change. But:

You might find that writing these using pytest fixtures makes the code nicer, and is easier to write
 
Here, you'd 
- add `@pytest.fixture` to this function, 
- rename it to `coords`, 
- and then any test function you wanted access to its result you'd add its name like `def test_units_in_data_and_coords(coords):`

You can have fixtures taking data from other fixtures, so you can still have the chain you're creating here, and you get parameterization for free!

Reply:
well, that was what I did before, but when I created this PR, I was told in https://github.com/pydata/xarray/pull/2956#discussion_r285430520 that data fixtures were adding complexity to the tests so I should either create the data in the tests themselves or use helper functions.

Reply:
I don't think it particularly matters whether you write fixtures or helper functions. My point was more about trying to keep test-specific inputs and assertions as close to each other as possible, so you can read a test without constantly referring to a helper function that defines the inputs.

Reply:
I hadn't seen that, sorry you're caught in the middle there @keewis. 

For this case, you can do it how you wish; I think @shoyer 's emphasis was on the locality rather than the format - so no benefit from `create_x` over an `x` fixture

(then we can all have a separate discussion about locality vs generality without holding this work up)

Reply:
should I then leave it as is or rewrite it so that the data is created in each individual test? None of these actually need all of data, dimensions and coordinates as subclass instances. Actually, they are almost the same as the ones in the original PR. I would be happy with both, though I do think rewriting may actually make the tests a little bit easier to understand.

Reply:
Assuming it's as straight swap between `create_X` and a fixture `X` (i.e. no change to the locality), I think they would be better as pytest fixtures. They're acceptable as either, though.

@shoyer what are your thoughts about the case at hand?

Reply:
What I meant was that at the moment I have a single 2d array subclass as data and both dimensions and an extra coordinate are also subclass instances. The dimensions (and the coordinate) only matter in the tests they are used in, so most of them could be in place created 1d arrays.

Reply:
I modified the tests to create the data in place.

Reply:
OK great! Cheers @keewis